src/crypto/hpke/mod.rs

```diff
@@ -0,0 +1,236 @@
+        const HELPER_ORIGIN: HelperOrigin<'static> = HelperOrigin("foo");
+        const SITE_ORIGIN: SiteOrigin<'static> = SiteOrigin("bar");
+
+        pub fn new(keys_per_epoch: usize, mut rng: R) -> Self {
+            Self {
+                registry: KeyRegistry::random(4, keys_per_epoch, &mut rng),
+                rng,
+                epoch: 0,
+            }
+        }
+
+        #[must_use]
+        pub fn seal(
+            &mut self,
+            key_id: KeyIdentifier,
+            match_key: XorReplicated,
+        ) -> MatchKeyEncryption {
+            let aad =
+                AssociatedData::new(key_id, self.epoch, &Self::HELPER_ORIGIN, &Self::SITE_ORIGIN)
+                    .unwrap();
+            let info = aad.to_bytes();
+            let mut plaintext = [0_u8; 16];
+            match_key.serialize(&mut plaintext).unwrap();
+            let pk_r = self.registry.public_key(self.epoch, key_id);
+
+            let (encap_key, tag) =
+                single_shot_seal_in_place_detached::<IpaAead, super::IpaKdf, super::IpaKem, _>(
+                    &OpModeS::Base,
+                    pk_r,
+                    &info,
+                    &mut plaintext,
+                    &[],
+                    &mut self.rng,
+                )
+                .unwrap();
+
+            MatchKeyEncryption {
+                encap_key: <[u8; 32]>::from(encap_key.to_bytes()),
+                ct: Box::new(plaintext),
+                tag: <[u8; 16]>::from(tag.to_bytes()),
+            }
+        }
+
+        pub fn open(
+            &self,
+            key_id: KeyIdentifier,
+            mut enc: MatchKeyEncryption,
+        ) -> Result<XorReplicated, BottomError> {
+            let aad =
+                AssociatedData::new(key_id, self.epoch, &Self::HELPER_ORIGIN, &Self::SITE_ORIGIN)
+                    .unwrap();
+            open_in_place(
+                &self.registry,
+                &enc.encap_key,
+                enc.ct.as_mut(),
+                &enc.tag,
+                &aad,
+            )?;
+
+            Ok(XorReplicated::deserialize(enc.ct.as_ref())?)
+        }
+
+        pub fn advance_epoch(&mut self) {
+            self.epoch += 1;
+        }
+    }
+
+    /// Make sure we obey the spec
+    #[test]
+    fn ipa_info_serialize() {
+        let aad =
+            AssociatedData::new(255, 32767, &HelperOrigin("foo"), &SiteOrigin("bar")).unwrap();
+        assert_eq!(
+            b"private-attributionfoo\0bar\0\xff\x7f\xff",
+            aad.to_bytes().as_ref()
+        );
+    }
+
+    #[test]
+    fn decrypt_happy_case() {
+        let rng = StdRng::from_seed([1_u8; 32]);
+        let mut suite = EncryptionSuite::new(1, rng);
+        let match_key = XorReplicated::new(u64::MAX, u64::MAX / 2);
+
+        let enc = suite.seal(0, match_key);
+        let r = suite.open(0, enc).unwrap();
+
+        assert_eq!(match_key, r);
+    }
+
+    #[test]
+    fn decrypt_wrong_aad() {
```

Comment:
Is this a case where you could do a proptest?

Reply:
I added some proptests already - what is the specific scenario you want to test here?

Reply:
You damage the ciphertext (and tag, which I generally don't treat as a distinct thing).  But you don't damage the info.  That's all.

Reply:
ah sure I can do that - randomly flip bits in info and validate ct cannot be opened

Reply:
I am going to randomly flip bits in random fields in `Info` struct. The API for `open` does not expose info byte string and I'll probably want to keep it that way